### DISCUSSION TITLE
Priority for stops

### DISCUSSION THREAD

**Discussion by TeodoraBogdan** [google/or-tools#3854] [Routing (and legacy CP) questions]
Hello!

I want to set priority for stops, for example high, medium and low priority. As expected the high priority stops will be visited first, then the medium and after the low ones. 
First I thought of creating 3 vectors with the stops, one for each priority level. And to use NextVar(stop)->RemoveValues(higherPriorityStops) to be sure that the stops will be visited in the order: high, medium and low.
But the problem is if the stops have time window. In this case, no solution may be found.
For example:
2 high priority stops, A,B with time window [8:00-9:00], [10:00-11:00]
1 medium priority stop C with time window [9:30-11:00]
and 3 low priority stops D,E,F without time window.
the start time is 08:00 and the route is round
With the RemoveValues() method, the solution would look like departure,A,B,C,D,E,F,departure (D,E,F in any order), but if the duration between A and B is 15 minutes, the vehicle will arrive too early at stop B, but I don't want to use wait times. So in this case I want the algorithm to maybe visit stop C (or even lower priority stops if needed), and only after to visit stop B => departure,A,C,B,D,E,F,departure (D,E,F in any order)

Another problem can be if I use weight constraint. And the vehicle has a maximum weight that it can carry, and for example the vehicle is full and it needs to visit a lower priority stop which may be a delivery stop, to unloads some goods, and only after this to go to a high priority one to collect the goods.

Is there a way to solve all this cases? (I am using C++)

Thanks!

**Comment by jmarca** [comment]
Hi,

Interesting problem and there is no one-size-fits-all solution.

Some ideas that might help:

If you track time in a dimension, you can set a soft upper bound of
some low time (or even 0:00 am) for the high priority nodes, 10:00am
for the medium, and no soft upper bound for the low priority nodes.
To avoid the penalties, the solver will put those as early as possible.

The problem with that approach is that you still might hit some low
priority nodes first.  So you might want to also set a soft *lower*
bound of some late time for the low priority nodes (say, 2pm).

Another idea is to set a requirement that some minimum number of high
priority nodes have to be visited before any medium priority nodes.
You can use a counting dimension for each node priority (high-count,
medium-count, low-count), and then say something like at a medium
priority node, the high count must be at least 90% of the high
priority nodes; and at low nodes the medium count must be 90% of the
medium priority nodes.  The problem there is of course if you have
multiple vehicles that will not work at all, as one might serve all
the high priority nodes, and the second will never visit enough to
trip the condition for visiting medium and low.  But with just one
vehicle it might work out okay.

Hope that gets you on a good path.

James


On Mon, Jul 10, 2023 at 11:52:19PM -0700, TeodoraBogdan wrote:
> Hello!
>
> I want to set priority for stops, for example high, medium and low priority. As expected the high priority stops will be visited first, then the medium and after the low ones.
> First I thought of creating 3 vectors with the stops, one for each priority level. And to use NextVar(stop)->RemoveValues(lowerPriorityStops) to be sure that the stops will be visited in the order: high, medium and low.
> But the problem is if the stops have time window. In this case, no solution may be found.
> For example:
> 2 high priority stops, A,B with time window [8:00-9:00], [10:00-11:00]
> 1 medium priority stop C with time window [9:30-11:00]
> and 3 low priority stops D,E,F without time window.
> the start time is 08:00 and the route is round
> With the RemoveValues() method, the solution would look like departure,A,B,C,D,E,F,departure (D,E,F in any order), but if the duration between A and B is 15 minutes, the vehicle will arrive too early at stop B, but I don't want to use wait times. So in this case I want the algorithm to maybe visit stop C (or even lower priority stops if needed), and only after to visit stop B => departure,A,C,B,D,E,F,departure (D,E,F in any order)
>
> Another problem can be if I use weight constraint. And the vehicle has a maximum weight that it can carry, and for example the vehicle is full and it needs to visit a lower priority stop which may be a delivery stop, to unloads some goods, and only after this to go to a high priority one to collect the goods.
>
> Is there a way to solve all this cases? (I am using C++)
>
> Thanks!
>
> --
> Reply to this email directly or view it on GitHub:
> https://github.com/google/or-tools/discussions/3854
> You are receiving this because you are subscribed to this thread.
>
> Message ID: ***@***.***>

--

James E. Marca
Activimetrics LLC
